### PR TITLE
feat(APIv2): RHICOMPL-3956 align offset with standard expectations

### PR DESCRIPTION
**Pull request by RoamingNoMaD**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/controllers/concerns/v2/metadata.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+        {
+          meta: {
+            total: opts[:total],
+            self.class::SEARCH => permitted_params[self.class::SEARCH],
+            tags: tags,
+            limit: pagination_limit,
+            offset: pagination_offset,
+            sort_by: permitted_params[:sort_by],
+            relationships: relationships_enabled?
```
Comment (skateman) [inline app/controllers/concerns/v2/metadata.rb:30]: This field can be removed, we won't use it in V2, at least not in its current form

File: app/controllers/concerns/v2/metadata.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+      end
+
+      def base_link_params
+        {
+          self.class::SEARCH => permitted_params[self.class::SEARCH],
+          include: permitted_params[:include],
+          limit: pagination_limit,
+          tags: permitted_params[:tags],
+          sort_by: permitted_params[:sort_by],
+          relationships: relationships_enabled?
```
Comment (skateman) [inline app/controllers/concerns/v2/metadata.rb:99]: Same here...

File: app/controllers/concerns/v2/pagination.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+    ParamType = ActionController::Parameters
+
+    included do
+      def pagination_limit
+        permitted_params[:limit] || 10
+      end
+
+      def pagination_offset
+        permitted_params[:offset] || 1
```
Comment (skateman) [inline app/controllers/concerns/v2/pagination.rb:16]: This has to be 0 by default.

**Review by skateman — CHANGES_REQUESTED** [review]
The V1 implementation wrongly uses the term offset, where it is actually representing page number. Offset means how many items are skipped from the beginning of the set when displaying the result, so the relationship between these two is `offset = (page_number - 1) * per_page`. The code should be adjusted accordingly...

File: app/controllers/concerns/v2/pagination.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      def pagination_limit
+        permitted_params[:limit] || 10
+      end
+
+      def pagination_offset
+        permitted_params[:offset] || 0
+      end
+
+      def page
+        pagination_offset + 1
```
Comment (skateman) [inline app/controllers/concerns/v2/pagination.rb:20]: This is definitely not good, the relationship between offset and page should be a multiplication of the page number with the per_page limit.

**Comment by skateman** [comment]
Something is off:
```
$ curl -k -H "Content-Type: application/json" -H "X-RH-IDENTITY: $RHI" -XGET "http://localhost:3000/api/compliance/v2/security_guides?limit=2" | jq '.data[].id'
"02ee36bf-d662-464c-938a-7ab025cf9db6"
"12cbb019-e776-4334-89ef-612b18c412c3"

$ curl -k -H "Content-Type: application/json" -H "X-RH-IDENTITY: $RHI" -XGET "http://localhost:3000/api/compliance/v2/security_guides?limit=2&offset=1" | jq '.data[].id'
"02ee36bf-d662-464c-938a-7ab025cf9db6"
"12cbb019-e776-4334-89ef-612b18c412c3"

$ curl -k -H "Content-Type: application/json" -H "X-RH-IDENTITY: $RHI" -XGET "http://localhost:3000/api/compliance/v2/security_guides?limit=2&offset=2" | jq '.data[].id'
"3a5db8e7-8ff2-41d9-a597-81dc762a760e"
"3fd0dbe6-7ad7-49c5-933e-0ff877890c15"

$ curl -k -H "Content-Type: application/json" -H "X-RH-IDENTITY: $RHI" -XGET "http://localhost:3000/api/compliance/v2/security_guides?limit=2&offset=3" | jq '.data[].id'
"3a5db8e7-8ff2-41d9-a597-81dc762a760e"
"3fd0dbe6-7ad7-49c5-933e-0ff877890c15"

curl -k -H "Content-Type: application/json" -H "X-RH-IDENTITY: $RHI" -XGET "http://localhost:3000/api/compliance/v2/security_guides?limit=2&offset=4" | jq '.data[].id'
"464e02d0-0220-4c8a-9371-2d821585f4eb"
"50cbb6ab-27ca-4393-9b08-eaecca41d657"
```

**Comment by RoamingNoMaD** [comment]
> Something is off:
> 
> ```
> $ curl -k -H "Content-Type: application/json" -H "X-RH-IDENTITY: $RHI" -XGET "http://localhost:3000/api/compliance/v2/security_guides?limit=2" | jq '.data[].id'
> "02ee36bf-d662-464c-938a-7ab025cf9db6"
> "12cbb019-e776-4334-89ef-612b18c412c3"
> 
> $ curl -k -H "Content-Type: application/json" -H "X-RH-IDENTITY: $RHI" -XGET "http://localhost:3000/api/compliance/v2/security_guides?limit=2&offset=1" | jq '.data[].id'
> "02ee36bf-d662-464c-938a-7ab025cf9db6"
> "12cbb019-e776-4334-89ef-612b18c412c3"
> 
> $ curl -k -H "Content-Type: application/json" -H "X-RH-IDENTITY: $RHI" -XGET "http://localhost:3000/api/compliance/v2/security_guides?limit=2&offset=2" | jq '.data[].id'
> "3a5db8e7-8ff2-41d9-a597-81dc762a760e"
> "3fd0dbe6-7ad7-49c5-933e-0ff877890c15"
> 
> $ curl -k -H "Content-Type: application/json" -H "X-RH-IDENTITY: $RHI" -XGET "http://localhost:3000/api/compliance/v2/security_guides?limit=2&offset=3" | jq '.data[].id'
> "3a5db8e7-8ff2-41d9-a597-81dc762a760e"
> "3fd0dbe6-7ad7-49c5-933e-0ff877890c15"
> 
> $ curl -k -H "Content-Type: application/json" -H "X-RH-IDENTITY: $RHI" -XGET "http://localhost:3000/api/compliance/v2/security_guides?limit=2&offset=4" | jq '.data[].id'
> "464e02d0-0220-4c8a-9371-2d821585f4eb"
> "50cbb6ab-27ca-4393-9b08-eaecca41d657"
> ```

**Comment by skateman** [comment]
Coverage issues seem relevant...

File: spec/controllers/v2/concerns/with_metadata_spec.rb [hunk 5]
```diff
@@ -16,7 +16,7 @@
 # ```
 #
 RSpec.shared_examples 'with metadata' do |*parents|
-  let(:meta_keys) { %w[total limit offset relationships] }
```
[the remark this answers is not in the document]
Comment (skateman) [reply spec/controllers/v2/concerns/with_metadata_spec.rb:19]: I agree, please at least move it to a separate commit.

File: app/controllers/concerns/v2/metadata.rb [hunk 6]
```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+module V2
+  # Sets up the metadata to be passed fo fast_jsonapi to be
```
Comment (romanblanco) [inline app/controllers/concerns/v2/metadata.rb:4]:
```suggestion
  # Sets up the metadata to be passed to fast_jsonapi to be
```

**Review by romanblanco — APPROVED** [review]
I have tested the PR and I'm receiving an expected outputs.

LGTM :+1:

**Review by skateman — APPROVED** [review]
The Seal of Approval

**Pull request merged**